File: solutions/python3/0042.py
class Solution:
    def trap(self, height: List[int]) -> int:
        ans = 0
        l = 0
        r = len(height) - 1
        leftMax = 0
        rightMax = 0

        while l < r:
            if height[l] < height[r]:
                if height[l] >= leftMax:
                    leftMax = height[l]
                else:
                    ans += (leftMax - height[l])
                l += 1
            else:
                if height[r] >= rightMax:
                    rightMax = height[r]
                else:
                    ans += (rightMax - height[r])
                r -= 1

        return ans

File: solutions/cpp/0042.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ret = 0;
        int l = 0;
        int r = height.size() - 1;
        int leftMax = 0;
        int rightMax = 0;

        while (l < r) {
            if (height[l] < height[r]) {
                if (height[l] >= leftMax)
                    leftMax = height[l];
                else
                    ret += (leftMax - height[l]);
                l++;
            } else {
                if (height[r] >= rightMax)
                    rightMax = height[r];
                else
                    ret += (rightMax - height[r]);
                r--;
            }
        }

        return ret;
    }
};